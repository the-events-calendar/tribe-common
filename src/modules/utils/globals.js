/**
 * @todo: handle globals in a better way
 */

export const get = ( key, defaultValue ) => window[ key ] || defaultValue;
export const google = () => get( 'google' );
export const wpApi = wp.api;
export const wpApiRequest = wp.apiRequest;
export const wpComponents = wp.components;
export const wpData = wp.data;
export const wpEditor = wp.blockEditor || wp.editor;
export const wpHooks = wp.hooks;

export const wpDataSelectCoreEditor = () => wpData.select( 'core/block-editor' ) ||
	wpData.select( 'core/editor' );

export const IconButton = wpComponents.Button || wpComponents.IconButton;

// Localized Config
export const config = () => get( 'tribe_editor_config', {} );

// Common
export const common = () => config().common || {};
export const adminUrl = () => common().adminUrl || '';
export const rest = () => common().rest || {};
export const restNonce = () => rest().nonce || {};
export const dateSettings = () => common().dateSettings || {};
export const editorConstants = () => common().constants || {};
export const list = () => ( {
	countries: common().countries || {},
	us_states: common().usStates || {},
} );

// TEC
export const tec = () => config().events || {};
export const editor = () => tec().editor || {};
export const settings = () => tec().settings || {};
export const mapsAPI = () => tec().googleMap || {};
export const priceSettings = () => tec().priceSettings || {};
export const tecDateSettings = () => tec().dateSettings || {};
export const timezoneHtml = () => tec().timezoneHTML || '';
export const defaultTimes = () => tec().defaultTimes || {};
export const timezone = () => tec().timeZone || {};

// PRO
export const pro = () => config().eventsPRO || {};
export const editorDefaults = () => pro().defaults || {};

// Tickets
export const tickets = () => config().tickets || {};

// Tickets Plus
export const ticketsPlus = () => config().ticketsPlus || {};
export const iacVars = () => ticketsPlus().iacVars || {};

// Post Objects
export const postObjects = () => config().post_objects || {};
export const post = () => config().post || {};

// Blocks
export const blocks = () => config().blocks || {};
