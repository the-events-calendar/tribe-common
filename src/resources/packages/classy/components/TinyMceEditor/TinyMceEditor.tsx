import * as React from 'react';
import { useEffect, useState } from 'react';
import { debounce } from 'lodash';
import '../../types/global.d.ts';

type TinyMceEditorProps = {
	content: string;
	onChange: ( value: string ) => void;
	id: string;
};

/**
 * Editor component that initializes and manages a TinyMCE editor.
 * @since TBD
 *
 * @param {TinyMceEditorProps} props - The component props.
 * @param {string} props.content The current content of the editor.
 * @param {(value: string) => void} props.onChange - Callback function to update the content.
 *
 * @return {JSX.Element} The rendered editor.
 */
export default function TinyMceEditor( { content, onChange, id }: TinyMceEditorProps ) {
	const [ value, setValue ] = useState( content );
	const [ , setEditor ] = useState< any >( null );

	useEffect( () => {
		// @ts-ignore - Defined by the `wp-tinymce` dependency required by the Classy package.
		if ( window.tinymce.get( id ) ) {
			// @ts-ignore - Defined by the `wp-tinymce` dependency required by the Classy package.
			window.wp.oldEditor.remove( id );
		}

		// @ts-ignore - Defined by the `wp-tinymce` dependency required by the Classy package.
		window.wp.oldEditor.initialize( id, {
			tinymce: {
				wpautop: true,
				toolbar1:
					'formatselect bold italic | bullist numlist blockquote | alignleft aligncenter alignright | link wp_more wp_adv',
				toolbar2:
					'strikethrough hr forecolor | pastetext removeformat charmap | outdent indent | undo redo | wp_help',
			},
			quicktags: true, // Show the "Visual / Text" tabs.
			mediaButtons: true, // Show the "Add media" button.
		} );

		// @ts-ignore - Defined by the `wp-tinymce` dependency required by the Classy package.
		const editor = window.tinymce.get( id );

		// Handle initialization of the editor.
		const onInit = () => {
			editor.on( 'NodeChange', debounce( triggerChangeIfDirty, 250 ) );
			setEditor( editor );
		};

		if ( editor.initialized ) {
			onInit();
		} else {
			editor.on( 'init', onInit );
		}

		// Cleanup function to remove event listeners
		return () => {
			if ( editor ) {
				editor.off( 'NodeChange', debounce( triggerChangeIfDirty, 250 ) );
			}
		};
	}, [ id ] );

	const triggerChangeIfDirty = () => {
		// @ts-ignore - Defined by the `wp-tinymce` dependency required by the Classy package.
		updateValues( window.wp.oldEditor.getContent( id ) );
	};

	const updateValues = ( newValue: string ) => {
		setValue( newValue );
		onChange( newValue );
	};

	return (
		<textarea
			className="classy-control-tinymce-editor wp-editor-area"
			id={ id }
			value={ value }
			onChange={ ( e ) => updateValues( e.target.value ) }
		/>
	);
}
