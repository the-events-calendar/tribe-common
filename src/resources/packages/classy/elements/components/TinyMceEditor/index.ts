export {default as TinyMceEditor} from './TinyMceEditor';
