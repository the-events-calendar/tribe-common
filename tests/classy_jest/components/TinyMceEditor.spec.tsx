import * as React from 'react';
import { render } from '@testing-library/react';
import { beforeEach, afterEach, describe, expect, it, jest } from '@jest/globals';
import TinyMceEditor from '../../../src/resources/packages/classy/components/TinyMceEditor/TinyMceEditor';
import { userEvent } from '@testing-library/user-event';

declare global {
	interface Window {
		wp: {
			oldEditor: {
				initialize: jest.Mock;
				getContent: jest.Mock;
				remove: jest.Mock;
			};
		};
		tinymce: {
			get: ( id: string ) => {
				on: jest.Mock;
			};
		};
	}
}

describe( 'TinyMceEditor Component', () => {
	const defaultProps = {
		content: '<p>Initial content</p>',
		onChange: jest.fn(),
		id: 'editor-id',
	};

	beforeEach( () => {
		jest.resetAllMocks();
	} );

	afterEach( () => {
		jest.resetAllMocks();
	} );

	it( 'renders correctly with default props', () => {
		const { container } = render( <TinyMceEditor { ...defaultProps } /> );

		expect( container.firstChild ).toMatchSnapshot();
	} );

	it( 'initializes the editor on mount', () => {
		render( <TinyMceEditor { ...defaultProps } /> );

		expect( global.window.wp.oldEditor.initialize ).toHaveBeenCalledWith( defaultProps.id, expect.any( Object ) );
	} );

	it( 'calls onChange when content changes', async () => {
		const user = userEvent.setup();
		const { container } = render( <TinyMceEditor { ...defaultProps } /> );

		const newEditorContent = '<p>New content</p>';
		const editor = container.querySelector( 'textarea.classy-control-tinymce-editor' ) as Element;

		expect( editor ).not.toBeNull();

		await user.clear( editor );
		await user.type( editor, newEditorContent );

		expect( defaultProps.onChange ).toHaveBeenCalledWith( newEditorContent );
	} );
} );
