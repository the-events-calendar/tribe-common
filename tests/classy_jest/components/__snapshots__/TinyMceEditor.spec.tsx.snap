// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TinyMceEditor Component renders correctly with default props 1`] = `
<textarea
  class="classy-control-tinymce-editor wp-editor-area"
  id="editor-id"
>
  &lt;p&gt;Initial content&lt;/p&gt;
</textarea>
`;
